### PR TITLE
MVP for localhost: Add voting functionality to CLI

### LINKED ISSUE #863: Uma Voting CLI Tool: Commit, Reveal
Project: **Build User Facing CLI tool with Two Interfaces**

1. Command line menu (for less technical users)
- Helpful tech: [`inquirer.js`](https://github.com/SBoudrias/Inquirer.js) 
2.  Traditional command line interface 
- Helpful tech: [`commander.js`](https://github.com/tj/commander.js) 

Actions will be roughly split into two categories:
- `wallet`: manages Ethereum (and therefore UMA) account creation, recovery, and displaying of token balances
- `voting`: enables user to view list of pending price requests they can vote on, commit and reveal votes
___

This issue specifically addresses the `voting` module which provides functionality for voting on pending price requests and dealing with the aftermath

### LINKED ISSUE #864: Uma Voting CLI Tool: Integrate into MetaMask

### LINKED ISSUE #878: Uma Voting CLI: Enable user to claim rewards
User needs way to view outstanding rewards and claim rewards

### LINKED ISSUE #879: Uma Voting CLI: Show time remaining in current round
Assumption: Rounds start and end on days (precisely)

Commit and Reveal phase each are one day long

Can use this information to determine when the next Phase starts and when the overall round ends

Display this in vote `info` method

### LINKED ISSUE #902: Uma Voting CLI: Work with 2 Key Contract
Allow user to pass in address of designated voting contract and vote via proxy

### PR DESCRIPTION
**Pull request by nicholaspai**
**User displays information about round, including pending price requests and rewards**
Screen Shot 2020-02-06 at 11 24 00

___

**User can view current price requests to vote on in the COMMIT phase**
-User will be prompted first to select which prices to vote on and then can enter prices one by one. Votes will be committed automatically in the optimal number of `batchCommits`

_Output of `Pending Commit or Reveal Requests` (there are many)_
Screen Shot 2020-02-05 at 13 03 48
_Output of `Commit`, prompting user to select prices to vote on_
Screen Shot 2020-02-05 at 13 04 16
_User manually entering prices_
Screen Shot 2020-02-05 at 13 04 34
_Transaction receipts and salts_
Screen Shot 2020-02-05 at 13 04 43
___

**User can view commited votes to reveal in the REVEAL phase**
-User will be prompted to select which votes to reveal. Reveals will be revealed automatically in the optimal number of `batchReveals`

_Output of `Pending Commit or Reveal Requests` (only reveal requests are shown for this phase)_
Screen Shot 2020-02-05 at 13 10 11
_Output of `Reveal`, prompting user to select votes to reveal_
Screen Shot 2020-02-05 at 13 10 34
_Transaction receipts_
Screen Shot 2020-02-05 at 13 10 41
___

**User can now retrieve rewards**
_The user will select which round ID of rewards to retrieve_
Screen Shot 2020-02-06 at 11 27 02
_Rewards will be retrieved in optimum batches_
Screen Shot 2020-02-06 at 11 27 12
___

**COLORS SHOULD BE CHANGED** They are currently there to show that we can add colors, but I have no systematic color scheme yet

Resolves #863 
Resolves #879 
Resolves #878 
Resolves #864 
Resolves #902 

Signed-off-by: Nick Pai <npai.nyc@gmail.com>

### CODE REVIEW

**Comment by nicholaspai** [comment]
Added functionality for retrieving rewards, therefore changed this to a WIP Pull Request.

Will need to refactor the rewards logic and test it more rigorously

File: core/scripts/local/RequestOraclePrice.js [hunk 1]
```diff
@@ -105,7 +105,9 @@ const runRequestOraclePrice = async function(callback) {
       await run(finder, argv.identifier, argv.time);
     });
   } else {
-    await run(finder, argv.identifier, argv.time);
+    for (let i = 0; i < 10; i++) {
```
Comment (nicholaspai) [inline core/scripts/local/RequestOraclePrice.js]: I will have to change this back before merging into prod

**Comment by nicholaspai** [comment]
@mrice32 @ptare @chrismaree 

Made one small change:
- Moved `core/scripts/cli/voting/constructCommitment and ./core/scripts/cli/voting/constructReveal` to `common/VotingUtils` because these functions are re-used by the AVS
- Accordingly, updated AVS (`core/scripts/Voting` to re-use `common/VotingUtils` and removed its `constructCommitment` and `constructReveal` methods)

File: core/scripts/cli/cli.js [hunk 2]
```diff
@@ -27,6 +28,8 @@ async function topMenu() {
 }
 
 async function run() {
+  // TODO: Should do a check here to detect if contracts are deployed,
```
Comment (chrismaree) [inline core/scripts/cli/cli.js]: Do you need to do this kind of check? Because the script runs in a truffle environment it will throw it's hands up if you don't have these artifacts. However I suppose if your web3 provider is pointing to a network that is not deployed this could cause an issue

Comment (nicholaspai) [reply core/scripts/cli/cli.js]: yeah right now it sort of just blows up if the contracts aren't deployed to the network. I do assume that the user has these artifacts though.

Comment (nicholaspai) [reply core/scripts/cli/cli.js]: again, super edge case, because in production the contracts would be deployed on mainnet, and presumably all test nets

File: core/scripts/cli/voting/batchCommitVotes.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
+const style = require("../textStyle");
```
Comment (mrice32) [inline core/scripts/cli/voting/batchCommitVotes.js]: Have you thought about combining batchCommitVotes.js, batchRevealVotes.js, and batchRetrieveRewards.js into one file?

This has a few advantages:
- Combining similar code into the same file (they all fundamentally follow the same pattern).
- Naming the functions inside the file, so it's easier to tell what the function does without having to look back at the filename.
- Reducing the overall number of files a reader needs to look through to understand the flow of data in the voting module.

There are also good reasons to keep them separate, so if you think they should stay separate, that's cool too.

File: core/scripts/cli/voting/retrieveRewards.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
+const getDefaultAccount = require("../wallet/getDefaultAccount");
+const style = require("../textStyle");
+const batchRetrieveRewards = require("./batchRetrieveRewards");
+const getResolvedVotesByRoundId = require("./getResolvedVotesByRoundId");
+const inquirer = require("inquirer");
+
+module.exports = async (web3, voting) => {
```
Comment (mrice32) [inline core/scripts/cli/voting/retrieveRewards.js]: nit: slightly prefer naming functions and exporting them at the bottom. I think it just gives the reader a little more in-built documentation about what's going on.

File: core/scripts/cli/voting/getResolvedVotesByRoundId.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
+const style = require("../textStyle");
+
+// Given a list of revealed votes for a user, return
```
Comment (mrice32) [inline core/scripts/cli/voting/getResolvedVotesByRoundId.js]: Awesome comment here. Would you mind adding a short (doesn't have to be this detailed) comment above all of the methods you export in other files?

File: core/scripts/cli/voting/batchCommitVotes.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+const { BATCH_MAX_COMMITS } = require("../../../../common/Constants");
+
+module.exports = async (newCommitments, votingContract, account) => {
```
Comment (mrice32) [inline core/scripts/cli/voting/batchCommitVotes.js]: Note: this comment may be worth a TODO rather than you doing it in this PR (up to you).

It seems like there's a decent amount of repeated functionality here and in the AVS. Have you thought about moving the batching functionality up to a common folder, like you did the `constructCommitment` function? I was thinking you could just have one big function that you pass all the commits (identifier, time, and price you want to vote on), and it would construct and batch them, returning to you one (possibly large) data structure containing the batches, receipts, salts, etc. This may simplify the logic here and allow all the low-level batching and encryption stuff to happen in one place to isolate that complexity.

This also might allow us to isolate and write automated tests for that code without having to touch our CLI frontend.

File: core/scripts/cli/voting/batchCommitVotes.js (hunk 3 above)
Comment (nicholaspai) [reply core/scripts/cli/voting/batchCommitVotes.js]: So  I don't think that these methods can be combined into one because they are different enough

But I do agree they can be and should be combined into one file--in fact it can also be in the `common` file for use in other scripts

Comment (mrice32) [reply core/scripts/cli/voting/batchCommitVotes.js]: Sorry if I miscommunicated -- yes, just the same file, I don't think it makes sense to make them all the same function or anything :)

File: core/scripts/cli/voting/filterRequestsByRound.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        const request = chronologicalPriceRequests[i];
+        filteredRequests.push({
+          identifier: request.identifier,
+          time: request.time
+        });
+      }
+    } else {
+      // Only display committed votes during the reveal phase (i.e.
+      // if an encrypted message exists for the identifier-timestamp)
+      for (let i = 0; i < chronologicalPriceRequests.length; i++) {
```
Comment (mrice32) [inline core/scripts/cli/voting/filterRequestsByRound.js]: nit: this one is more optional (may be cleaner the way you have it), but wanted to point out a slightly different way of doing this:
```
filteredRequests = chronologicalPriceRequests.filter(request => {
  const topicHash = computeTopicHash(request, roundId);
  const encryptedCommit = await votingContract.getMessage(account, topicHash, { from: account });
  return encryptedCommit;
});
```

File: core/scripts/cli/voting/getResolvedVotesByRoundId.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+      continue;
+    }
+  }
+
+  // Create a mapping of round IDs to total rewards per round
+  let roundIds = [];
+  Object.keys(resolvedVotesByRoundId).forEach(id => {
+    let totalRewardsInRound = 0;
+    resolvedVotesByRoundId[id].forEach(reward => {
+      totalRewardsInRound += parseInt(reward.potentialRewards);
```
Comment (mrice32) [inline core/scripts/cli/voting/getResolvedVotesByRoundId.js]: nit: might be better to keep these as BNs, so you don't lose precision. You can always use the `fromWei()` function to do the division by `10e18`, and you can `.toString()` them whenever you need to print them out.

File: core/scripts/cli/voting/displayStatus.js [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    });
+    console.table(requestsTable);
+  }
+
+  // Display rewards to be retrieved in a table
+  console.log(`${style.success(`- Voting Rewards Available`)}:`);
+  if (rewards.roundIds.length > 0) {
+    const rewardsTable = [];
+    Object.keys(rewards.resolvedVotesByRoundId).forEach(roundId => {
+      rewards.resolvedVotesByRoundId[roundId].forEach(_reward => {
```
Comment (mrice32) [inline core/scripts/cli/voting/displayStatus.js]: Not sure if this is cleaner, but I thought I might throw another option out there for how to do this loop:
```
const reducer = (accumulator, currentValue) => accumulator + currentValue;
const rewardsTable = Object.values(rewards.resolvedVotesByRoundId).reduce(reducer).map(reward => {
 return {
      "Round ID": _reward.roundId,
      Name: _reward.name,
      "Reward Tokens": web3.utils.fromWei(_reward.potentialRewards)
    };
});
```
^After writing that out, it looks a little nasty too. I feel like this one just has to be kinda complicated :/

File: core/scripts/cli/voting/displayStatus.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        rewardsTable.push({
+          "Round ID": _reward.roundId,
+          Name: _reward.name,
```
Comment (mrice32) [inline core/scripts/cli/voting/displayStatus.js]: nit: why is this capitalized?

File: core/scripts/cli/voting/votePhaseTiming.js [hunk 11]
```diff
@@ -0,0 +1,20 @@
+const { VotePhasesEnum } = require("../../../../common/Enums");
+const moment = require("moment");
+const SECONDS_IN_DAY = 86400;
+const MINUTES_IN_HOUR = 60;
+
+module.exports = (contractTime, roundPhase) => {
```
Comment (mrice32) [inline core/scripts/cli/voting/votePhaseTiming.js]: maybe add a comment noting what type you're expecting for `contractType` and `roundPhase`? Are they both strings? Would this work if they were BNs?

File: core/scripts/cli/voting/revealVotes.js [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+const { constructReveal } = require("../../../../common/VotingUtils");
+const batchRevealVotes = require("./batchRevealVotes");
+
+module.exports = async (web3, voting) => {
+  style.spinnerReadingContracts.start();
+  const pendingRequests = await voting.getPendingRequests();
+  const roundId = await voting.getCurrentRoundId();
+  const roundPhase = await voting.getVotePhase();
+  const account = await getDefaultAccount(web3);
+  const filteredRequests = await filterRequests(pendingRequests, account, roundId, roundPhase, voting);
```
Comment (mrice32) [inline core/scripts/cli/voting/revealVotes.js]: I really like this all-in-one filter method. Very nice.

File: core/scripts/cli/voting/retrieveRewards.js (hunk 4 above)
Comment (nicholaspai) [reply core/scripts/cli/voting/retrieveRewards.js]: I agree, I gave a name to every method

File: core/scripts/cli/voting/getResolvedVotesByRoundId.js (hunk 5 above)
Comment (nicholaspai) [reply core/scripts/cli/voting/getResolvedVotesByRoundId.js]: added comments to every single method in `cli/`

File: core/scripts/cli/voting/batchCommitVotes.js (hunk 6 above)
Comment (nicholaspai) [reply core/scripts/cli/voting/batchCommitVotes.js]: done

File: core/scripts/cli/voting/commitVotes.js [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+          type: "number",
+          name: "price",
+          default: 0,
+          message: style.instruction(`Price for ${selections[i]}:`),
+          validate: value => value >= 0 || "Price must be positive"
+        });
+
+        // Look up raw request data from checkbox value
+        let selectedRequest;
+        for (let j = 0; j < filteredRequests.length; j++) {
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply core/scripts/cli/voting/commitVotes.js]: yup that works

File: core/scripts/cli/voting/commitVotes.js [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        style.instruction(
+          `Next, enter prices for the selected requests. Prices must be positive numbers, invalid input will default to 0.`
+        )
+      );
+
+      const newCommitments = [];
+      const failures = [];
+
+      // Prompt user to enter a price per vote construct commitments for the votes
+      const selections = checkbox["requestsCheckbox"];
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply core/scripts/cli/voting/commitVotes.js]: You bringing this up made me realize I can remove the whole 

```js
const selectedRequest = filteredRequests.find(request => request.value === selections[i])
```

So `inquirer.js` works as follows:

```js
const checkbox = await inquirer.prompt({
      type: "checkbox",
      name: "requestsCheckbox",
      message: `After selecting which requests you want to vote on, you will be prompted to manually enter in a price for each request. You can change these votes later.`,
      choices: filteredRequests
    });
```

`filteredRequests` is an array of objects that must contain a `name` and/or `value` parameter.  If there is a `name` parameter, then the `name` is displayed in the menu that the user is shown to select requests. Then `const selections = checkbox["requestsCheckbox"]` is set as an array of the `value` parameters.

So, I modified the `filteredRequests` that is passed to `inquirer.prompt` as follows:

```js
for (let i = 0; i < filteredRequests.length; i++) {
      let request = filteredRequests[i];
      request.name = `${web3.utils.hexToUtf8(request.identifier)} @ ${style.formatSecondsToUtc(
        parseInt(request.time)
      )}`;
      request.value = { identifier: request.identifier, time: request.time }
    }
```

File: core/scripts/cli/voting/commitVotes.js [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+      for (let i = 0; i < selections.length; i++) {
+        // Prompt user to enter a price per vote and commit the votes
+        const priceInput = await inquirer.prompt({
+          type: "number",
+          name: "price",
+          default: 0,
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply core/scripts/cli/voting/commitVotes.js]: yeah without the default, `inquirer` doesn't play nicely. It will display an error message and then put a `NaN` in the user's type box which for some reason I cannot backspace/delete

File: core/scripts/cli/voting/revealVotes.js [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    if (checkbox["requestsCheckbox"]) {
+      const newReveals = [];
+      const failures = [];
+
+      // Prompt user to enter a price per vote construct commitments for the votes
+      const selections = checkbox["requestsCheckbox"];
+      for (let i = 0; i < selections.length; i++) {
+        // Look up raw request data from checkbox value
+        let selectedRequest;
+        for (let j = 0; j < filteredRequests.length; j++) {
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply core/scripts/cli/voting/revealVotes.js]: no longer need to do this, see comment above

File: core/scripts/cli/voting/filterRequestsByRound.js [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    const chronologicalPriceRequests = pendingRequests.sort((a, b) => {
+      return parseInt(a.time) - parseInt(b.time);
+    });
+
+    // Depending on the round phase, determine which requests to display
+    if (roundPhase.toString() === VotePhasesEnum.COMMIT) {
+      // Display all requests during commit phase even if
+      // user has already committed a vote, for they
+      // might want to change it
+      for (let i = 0; i < chronologicalPriceRequests.length; i++) {
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply core/scripts/cli/voting/filterRequestsByRound.js]: yup

File: core/scripts/cli/voting/filterRequestsByRound.js (hunk 7 above)
Comment (nicholaspai) [reply core/scripts/cli/voting/filterRequestsByRound.js]: yeah this is more the more functional way to do it and saves a couple lines, I implemented it

File: core/scripts/cli/voting/getResolvedVotesByRoundId.js (hunk 8 above)
Comment (nicholaspai) [reply core/scripts/cli/voting/getResolvedVotesByRoundId.js]: good idea done

File: core/scripts/cli/voting/displayStatus.js (hunk 9 above)
Comment (nicholaspai) [reply core/scripts/cli/voting/displayStatus.js]: yea i think the map reduce is cleaner but is less clear

File: core/scripts/cli/voting/displayStatus.js (hunk 10 above)
Comment (nicholaspai) [reply core/scripts/cli/voting/displayStatus.js]: because all of the labels in shown to user are capitalized, I put this in double quotation marks to make it clearer that it is a table. This is the example output

```
** Your voting status **
  - Current round ID: 9157
  - Current round phase: Reveal
  - Round Inflation: 0.05 %
  - Round GAT: 5 %
  - Contract time: February 23rd 2020, 8:46:35 pm
  - Time until Commit phase: 3 hours, 13 minutes
  - Time until next voting round: 3 hours, 13 minutes
  - Pending Reveal Requests:
  ┌─────────┬──────────────┬───────────────────────────────────┐
  │ (index) │  Identifier  │               Time                │
  ├─────────┼──────────────┼───────────────────────────────────┤
  │    0    │ 'population' │ 'September 30th 2019, 7:20:31 pm' │
  │    1    │ 'population' │ 'September 30th 2019, 7:20:51 pm' │
  └─────────┴──────────────┴───────────────────────────────────┘
  - Voting Rewards Available:
  ┌─────────┬──────────┬────────────────────────────────────────────────┬───────────────┐
  │ (index) │ Round ID │                      Name                      │ Reward Tokens │
  ├─────────┼──────────┼────────────────────────────────────────────────┼───────────────┤
  │    0    │  '9152'  │  'bezos hairs @ August 15th 2019, 7:07:49 am'  │    '50050'    │
  │    1    │  '9156'  │ 'population @ September 23rd 2019, 8:40:51 pm' │    '50050'    │
  └─────────┴──────────┴────────────────────────────────────────────────┴───────────────┘
  
  
```

File: core/scripts/cli/voting/votePhaseTiming.js (hunk 11 above)
Comment (nicholaspai) [reply core/scripts/cli/voting/votePhaseTiming.js]: ah yeah both strings

File: core/scripts/cli/voting/displayStatus.js (hunk 10 above)
Comment (mrice32) [reply core/scripts/cli/voting/displayStatus.js]: Gotcha. Makes sense.

File: core/scripts/cli/voting/displayStatus.js (hunk 9 above)
Comment (mrice32) [reply core/scripts/cli/voting/displayStatus.js]: Nonetheless, I think you can use `Object.values()` so you don't have to re-access the object in the as you do on this line: `rewards.resolvedVotesByRoundId[roundId]`

Comment (nicholaspai) [reply core/scripts/cli/voting/displayStatus.js]: this doesn't work because this is the shape of `rewards.resolvedVotesByRoundId`

```
 {
    '9152': [
      {
        price: '122200000000000000000',
        name: 'bezos hairs @ August 15th 2019, 7:07:49 am',
        identifier: '0x62657a6f73206861697273000000000000000000000000000000000000000000',
        time: '1565852869',
        roundId: '9152',
        potentialRewards: '50050000000000000000000'
      }
    ],
    '9156': [
      {
        price: '0',
        name: 'population @ September 23rd 2019, 8:40:51 pm',
        identifier: '0x706f70756c6174696f6e00000000000000000000000000000000000000000000',
        time: '1569271251',
        roundId: '9156',
        potentialRewards: '50050000000000000000000'
      }
    ]
  }
```

Comment (nicholaspai) [reply core/scripts/cli/voting/displayStatus.js]: This reducer works however

```js
const reducer = (accumulator, currentValue) => accumulator.concat(currentValue);
```

File: core/scripts/cli/voting/displayStatus.js (hunk 10 above)
Comment (nicholaspai) [reply core/scripts/cli/voting/displayStatus.js]: I changed these to lower case, because Prettier complains if I leave the object as:

```
{
          "Round ID": _reward.roundId,
          "Name": _reward.name,
}
```

Prettier wants the "Name" to be without quotation marks, which I think looks weird. So I made all object keys in these tables lowercase

**Comment by nicholaspai** [comment]
Any last comments before we can merge? @mrice32 @chrismaree @ptare

**Comment by nicholaspai** [comment]
Can now vote with two key contract via the environment variable `TWO_KEY_ADDRESS` (instructions in `README.md`) and commented out code in `cli/voting/displayStatus.js` that works with deployed mainnet contract--namely, the deployed contracts do not publicly reveal inflation or gat percentages

Also now displays resolved prices that user has voted on

new example `Voting > Info`

Screen Shot 2020-02-10 at 15 07 27
